README
TFTP Theft is a tool which allows one to quickly scan/bruteforce a tftp server
for files and download them instantly.

thief.py - downloads files based on a pattern or list of files
finder.py - looks for tftp servers on a target ip address range

Author: sandro gauci <sandro@enablesecurity.com>


Changelog
v0.1 (20100526)

First version out

finder.py
#!/usr/bin/env python
# copyrighted by sandrogauci <sandro@enablesecurity.com> 2010
# 

import socket
import select 
import sys
import datetime
import random
from lib.tftplib import tftp, tftpstruct
from lib.iphelper import ip4range
import random
import os
from lib.common import __LICENSE__, __version__
from time import sleep
import logging
from optparse import OptionParser

DESC =    """

   Finder.py finds tftp servers quickly
    """


def getargs():
    usage = "usage: %prog [options] target\r\n"
    usage += "examples:\r\n"
    usage += "%prog 10.0.0.1/18 192.168.2.1/24\r\n"
    usage += "%prog -p6969  10.0.0.1/24\r\n"
    parser = OptionParser(usage, version="%prog v"+str(__version__)+DESC+__LICENSE__)
    parser.add_option("--port", '-p', dest="port", default=69, type="int",
                      help="Destination port")
    (options, args) = parser.parse_args()
    if len(args) == 0:
        parser.error("Please pass some ip ranges")
    return (options, args)

def main():
    log = logging.getLogger(__name__)
    log.setLevel(logging.INFO)
    options,args = getargs()
    ipranges = ip4range(*args)
    port = options.port
    fn = str(random.randrange(1000,9999)) + '.xml'
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    s.settimeout(5)
    t = tftp()
    start_time = datetime.datetime.now()
    finito = False
    data = t.makerrq(fn)
    while 1:
        r,w,e = select.select([s],[],[],0.1)
        if len(r) > 0:        
            for s in r:
                try:
                    recv = s.recvfrom(1024)
                except socket.error,e:
                    print e
                    continue            
                buff = recv[0]
                ipaddr = recv[1]
                try:
                    response = tftpstruct.parse(buff)
                except Exception,e:
                    print e
                log.info("IP: %s:%s responded: %s" % (ipaddr[0],ipaddr[1],response))
        else:
            if finito:
                break
            if not finito:
                try:
                    dst = ipranges.next(),port
                except StopIteration:
                    finito = True
                    continue
                try:
                    log.debug('sending to %s:%s' % dst)
                    s.sendto(data,dst)
                except socket.error, err:
                    log.warn('Socket error: %s %s' % (dst,err))
                    
    
    log.info("Duration: %s" %  (datetime.datetime.now() - start_time))

if __name__ == "__main__":
    logging.basicConfig()
    main()

lib/common.py
#!/usr/bin/env python

__version__ = '0.1'

__LICENSE__ = """

Copyright (c) 2010, Sandro Gauci <sandro@enablesecurity.com>
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
    * Neither the name of the <ORGANIZATION> nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""


lib/tftplib.py
"""
the Trivial FTP protocol
"""
from construct import *
from construct.text import *

op_code = Enum(UBInt16("operation"),
               RRQ = 1,
               WRQ = 2,
               DATA = 3,
               ACK = 4,
               ERROR = 5
              )

source_file = CString("source_file")
transfer_mode = CString("transfer_mode")
fileoperations = Struct(None,source_file,transfer_mode)
erroroperation = Struct(None,
                        Enum(UBInt16("errorcode"),
                             Unknown = 0,
                             FileNotFound = 1,
                             AccessViolation = 2,
                             DiskFullOrAllocationExceeded = 3,
                             IllegalTftpOperation = 4,
                             UnknownTransferId = 5,
                             FileAlreadyExists = 6,
                             NoSuchUser = 7,
                            ),
                        CString("ErrMsg"))
datapacket = Struct(None,
                    UBInt16("block"),
                    StringAdapter(GreedyRange(Char("data")))
                   )
ackpacket = Struct(None,
                   UBInt16("block"))
tftpstruct = Struct("command",
                      op_code,
                      Switch("data", lambda ctx: ctx["operation"],
                             {
                                'RRQ' : fileoperations,
                                'WRQ' : fileoperations,
                                'DATA' : datapacket,
                                'ACK' : ackpacket,
                                'ERROR' : erroroperation
                             }
                            )
                      )

class tftp:
    def __init__(self,buff=None):
        pass
    
    def parse(self,buff=None):
      try:
         return tftpstruct.parse(buff)
      except lib.construct.core.RangeError:
         pass
    
    def makerrq(self,filename,mode='octet'):
        rrq = Container()
        rrq.operation = 'RRQ'
        rrq.data = Container()
        rrq.data.source_file = filename
        rrq.data.transfer_mode = mode
        return tftpstruct.build(rrq)
      
    def makeack(self,block):
         ack = Container()
         ack.operation = 'ACK'
         ack.data = Container()
         ack.data.block = block
         return tftpstruct.build(ack)
        


if __name__ == "__main__":
    cap1 = (
    "000143544c5345503030304432393041423836302e746c76006f6374657400"
    ).decode("hex")

    cap2 = (
    "0005000146696c65206e6f7420666f756e6400"
    ).decode("hex")
    
    cap3 = (
    "000300013c64657669636520207873693a747970653d2261786c3a58495050"
    "686f6e65222063746969643d2232312220757569643d227b64366438393933"
    "362d383162322d623736642d343162632d3337666665323137333139657d22"
    "3e0d0a3c66756c6c436f6e6669673e747275653c2f66756c6c436f6e666967"
    "3e0d0a3c64657669636550726f746f636f6c3e534343503c2f646576696365"
    "50726f746f636f6c3e0d0a3c7373685573657249643e3c2f73736855736572"
    "49643e0d0a3c73736850617373776f72643e3c2f73736850617373776f7264"
    "3e0d0a3c646576696365506f6f6c2020757569643d227b3162316239656236"
    "2d373830332d313164332d626466302d3030313038333032656164317d223e"
    "0d0a3c7265766572745072696f726974793e303c2f7265766572745072696f"
    "726974793e0d0a3c6e616d653e44656661756c743c2f6e616d653e0d0a3c64"
    "61746554696d6553657474696e672020757569643d227b3965633438353061"
    "2d373734382d313164332d626466302d3030313038333032656164317d223e"
    "0d0a3c6e616d653e434d4c6f63616c3c2f6e616d653e0d0a3c646174655465"
    "6d706c6174653e4d2f442f593c2f6461746554656d706c6174653e0d0a3c74"
    "696d655a6f6e653e477265656e77696368205374616e646172642054696d65"
    "3c2f74696d655a6f6e653e0d0a3c2f6461746554"
    ).decode("hex")
    
    cap4 = (
    "00040001123123"
    ).decode("hex")
    caps = [cap1,cap2,cap3,cap4]
    for cap in caps:
        obj = tftpstruct.parse(cap)
        if obj.operation=='ERROR':            
            x = obj.data
            print x.errorcode
    
    t = tftp()
    print t.makerrq('hello.txt')

lib/iphelper.py
#!/usr/bin/env python

# comments / license
import struct
import socket


def dottedQuadToNum(ip):
    "convert decimal dotted quad string to long integer"
    return struct.unpack('!L',socket.inet_aton(ip))[0]

def numToDottedQuad(n):
    "convert long int to dotted quad string"
    return socket.inet_ntoa(struct.pack('!L',n))
                                                                                                             

def ip4range(*args):
    """takes a list ip ranges as an argument and returns a generator object                                                                                                         
    examples accepted:                                                                                                           
        1.1.1.1/16
        1.1.1.1-25
        1.1.1.1-1.1.2.255
        www.google.com/24
                                                                                                     
    """
    for arg in args:                                              
        r = getranges(arg)
        if r is None:                                                                                                   
            continue                                                                                                               
        startip,endip = r
        curip = startip
        while curip <= endip:
            yield(numToDottedQuad(curip))
            curip += 1                                                                                                    

def calcmasksize(ipstring):                                                                                                                         
    """given an ipstring this function returns the number of IPs"""
    tmp = getmaskranges(ipstring)                                                                
    if tmp is not None:
        naddr1,naddr2 = tmp
        size = naddr2 - naddr1 + 1
        return size                                                                                                                

def getranges(ipstring):
    """
    Tries to get a range of ip addresses based on a string
    supports the following formats:
     -  1.2.3.4-1.2.4.5
     -  1-2.3.4.5-9
     -  www.test.com/24
     -  1.2.3.4/24
     -  hostname
     
    >>> getranges('127.0.0.1')
    (2130706433, 2130706433)
    
    >>> getranges('1-2.3.4.5-9')
    (16974853, 33752073)
    
    >>> getranges('localhost/24')
    (2130706433L, 2130706686L)
    
    >>> getranges('127.0.0.1-127.0.0.255')
    (2130706433, 2130706687)
    
    >>> getranges('127.0.0.1/24')
    (2130706433L, 2130706686L)
    
    >>> getranges('localhost')
    (2130706433, 2130706433)
    
    >>> getranges('127.0.0.*')
    (2130706432, 2130706687)
    """
    import re
    from socket import gethostbyname                                                                                                                  
    if re.match(                                                                                                     
        '^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}-\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$',
        ipstring                                                                                                             
        ):
        naddr1,naddr2 = map(dottedQuadToNum,ipstring.split('-'))
    elif re.match(                                                                                                         
        '^(\d{1,3}(-\d{1,3})*)\.(\*|\d{1,3}(-\d{1,3})*)\.(\*|\d{1,3}(-\d{1,3})*)\.(\*|\d{1,3}(-\d{1,3})*)$',
        ipstring
        ):                                                                                                     
        naddr1,naddr2 = map(dottedQuadToNum,getranges2(ipstring))
    elif re.match(                                                                                                                  
        '^.*?\/\d{,2}$',
        ipstring
        ):
        r = getmaskranges(ipstring)
        if r is None:                                                                                                     
            return
        naddr1,naddr2 = r                                                                                                       
    else:
        # we attempt to resolve the host
        try:                                                                                                            
            naddr1 = dottedQuadToNum(gethostbyname(ipstring))
            naddr2 = naddr1
        except socket.error:
            ("sip",'Could not resolve %s' % ipstring)
            return                                                                                                               
    return((naddr1,naddr2))

def getranges2(ipstring):                                                                                                  
    _tmp = ipstring.split('.')                                                                                                 
    if len(_tmp) != 4:
        raise ValueError, "needs to be a Quad dotted ip"
    _tmp2 = map(lambda x: x.split('-'),_tmp)                                                                                                            
    startip = list()
    endip = list()
    for dot in _tmp2:
        if dot[0] == '*':                                              
            startip.append('0')
            endip.append('255')
        elif len(dot) == 1:
            startip.append(dot[0])                                                                                                   
            endip.append(dot[0])
        elif len(dot) == 2:
            startip.append(dot[0])
            endip.append(dot[1])
    naddr1 = '.'.join(startip)                                                                                                               
    naddr2 = '.'.join(endip)
    return(naddr1,naddr2)


def getmaskranges(ipstring):
    """gets the lower and upper ip address when given a mask
    
    >>> getmaskranges('127.0.0.1/32')
    (2130706433, 2130706433)
    
    >>> getmaskranges('127.0.0.1/24')
    (2130706433L, 2130706686L)
    
    >>> getmaskranges('1.1.1.1/24')
    (16843009L, 16843262L)
    """
    import re
    from socket import gethostbyname
    addr,mask = ipstring.rsplit('/',1)
    if not re.match('^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$',addr):
        try:            
            addr = gethostbyname(addr)
        except socket.error:
            return

    naddr = dottedQuadToNum(addr)
    masklen = int(mask)
    if masklen==32:
        return(naddr,naddr)
    if not 0 <= masklen <= 32:
        raise ValueError
    naddr1 = naddr & (((1<<masklen)-1)<<(32-masklen)) + 1
    naddr2 = naddr1 + (1<<(32-masklen)) - 3
    return (naddr1,naddr2)


def getlocalip():
    try:
        myip = socket.gethostbyname(socket.gethostname())
    except socket.error:
        #log.error("could not retrieve the local ip address - please specify via the -x option")
        return
    return myip


def mysendto(sock,data,dst):
    """A sendto replacement which splits up large packets into smaller 8192b packets

    Arguments:
    sock -- a socket object
    data -- the data arg to pass to sock.sendto
    dst -- the destination (ip,port) tuple to pass to sock.sendto

    """
    sock.connect(dst)
    while data:
        #if socknode.nodetype == 'LocalNode':
            #bytes_sent = sock.sendto(data[:8192],dst)
            #data = data[bytes_sent:]
        bytes_sent = sock.send(data[:8192])
        data = data[bytes_sent:]



def bindto(bindingip,startport,s):
    """Binds to a specific port and port. If the port is not available, then increments
    until a free port is found.

    Arguments:
    bindingip -- a string containing an IP address
    startport -- the prefered port to bind to
    s -- socket object

    """
    localport = startport
    #devlog("sip","binding to %s:%s" % (bindingip,localport))
    while 1:
        if localport > 65535:
            ("sip","Could not bind to any port")
            return
        try:
            s.bind((bindingip,localport))
            break
        except socket.error:
            #devlog("sip","could not bind to %s:%s" % (bindingip,localport))
            localport += 1
    return localport,s


def scanlist(iprange,portranges,methods):
    """ a generator that iterates on a range of ip addresses and ports and methods
    """
    for ip in iter(iprange):
        for portrange in portranges:
            for port in portrange:
                for method in methods:
                    yield(ip,port,method)

if __name__ == "__main__":
    import doctest
    doctest.testmod()

thief.py
#!/usr/bin/env python
# copyrighted by sandrogauci <sandro@enablesecurity.com> 2010
# 

import random
import os
from time import sleep, time
from optparse import OptionParser
import logging
import socket
import select 
import sys
import datetime
from lib.tftplib import tftp, tftpstruct
from lib.bruteforcehelper import anotherxrange, getRange
from lib.common import __LICENSE__, __version__
import lib.construct

DESC =    """

   Thief.py downloads files from a tftp server quickly and efficiently
   """


def gen_range(template,myranges):
    for myrange in myranges:
        for i in myrange:            
            outstr = template % i
            yield(outstr)

def gen_tftpfilelist(fn):
    f = open(fn,'r')
    while 1:
        line = f.readline().strip()
        if len(line) == 0:
            f.close()
            break
        yield(line)

    
def assemblefiles(dumpfiles,sportmap):
    files = dict()
    for addresses in dumpfiles:
        transferdone = False
        tmpfile = str()
        for part in dumpfiles[addresses]:
            data = dumpfiles[addresses][part] 
            tmpfile += data
            if len(data) < 512:
                transferdone = True
        sport = addresses[1][1]
        fn = sportmap[sport]
        files[(fn,transferdone)]=tmpfile
    return files

def splitpartialfull(dumpfiles):
    partial = dict()
    full = dict()
    for addresses in dumpfiles:
        transferdone = False
        for part in dumpfiles[addresses]:
            data = dumpfiles[addresses][part]
            if len(data) < 512:
                transferdone = True
                break
        if transferdone:
            full[addresses] = dumpfiles[addresses]
        else:
            partial[addresses] = dumpfiles[addresses]
    return full,partial

def dumpnassemble(dumpfiles,sportmap,dstdir,partialdump=True):
    partialfiles = None
    if not partialdump:
        dumpfiles,partialfiles = splitpartialfull(dumpfiles)    
    files = assemblefiles(dumpfiles,sportmap)
    donefiles = list()
    for (fn,done) in files:
        if not os.path.exists(dstdir):
            os.mkdir(dstdir)
        data=files[(fn,done)]
        f=open(os.path.join(dstdir,fn),'wb')
        f.write(data)
        f.close()
        donefiles.append(fn)
    return partialfiles,donefiles
        

class mylist:
    def __init__(self,limit=20):
        self.limit = limit
        self.list = []
    
    def append(self,data):
        if len(self.list) > self.limit-1:
            last = self.list.pop(-1)
            if type(last) == socket.socket:
                last.close()
        self.list.append(data)

def getargs():
    usage = "usage: %prog [options] target\r\n"
    usage += "examples:\r\n"
    usage += "%prog 10.0.0.1\r\n"
    usage += "%prog -p6969  10.0.0.1\r\n"
    parser = OptionParser(usage, version="%prog v"+str(__version__)+DESC+__LICENSE__)
    parser.add_option('-p',"--port", dest="port", default=69, type="int",
                      help="Destination port")
    parser.add_option('--range','-r', dest="range",
                      help="Range of hex (default) or numeric filenames")
    parser.add_option('--rangetype','-T', dest="rangetype", default="hex",
                      type="choice", choices=["hex","num"],
                      help="Range of hex (default, useful for mac addresses) or num for numeric filenames")
    parser.add_option('--fntemplate',dest="fntemplate", default="SEP%012X.cnf.xml",
                      help="""When the --range option is used, by default it produces
                      filenames for SCCP configuration files using the template
                      SEP%012X.cnf.xml. You can modify this to reflect any template or
                      specify a template set in fntemplates.txt""")
    parser.add_option('--file','-f', dest="filenamesfile",
                      default=os.path.join('data','tftplist.txt'),
                      help="File containing a list of files to download")
    parser.add_option('-o','--output',dest="dstdir", default="download",
                      help="Output directory where the files will be downloaded")
    parser.add_option('--listtemplates','-l', dest="listtemplates",
                      default=False, action="store_true",
                      help="List known templates")
    (options, args) = parser.parse_args()
    if len(args) != 1:
        parser.error("Please pass just one destination tftp server name")
    return (options, args)

def getfntemplate(fntemplate):
    import csv
    res = fntemplate
    f=open(os.path.join('data','fntemplates.txt'),'r')
    reader=csv.reader(f,delimiter=":")
    for row in reader:
        if row[0] == fntemplate:
            res = row[1]
            break
    f.close()
    return res

def listtemplates():
    f=open(os.path.join('data','fntemplates.txt'),'r')
    reader=csv.reader(f,delimiter=":")
    res = list()
    for row in reader:
        if row[0] == fntemplate:
            res.append(row[0])
            break
    f.close()
    return res
    
    
def main():
    log = logging.getLogger(__name__)
    log.setLevel(logging.INFO)
    options,args = getargs()
    if options.listtemplates:
        print listtemplates()
    dst = (args[0],int(options.port))
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.settimeout(5)
    mydirtysocks = mylist(limit=100)
    mydirtysocks.append(s)
    t = tftp()
    if options.range:
        ranges = getRange(options.range,rangetype=options.rangetype)
        fntemplate = getfntemplate(options.fntemplate)
        try:
            fntemplate % 100
        except TypeError:
            log.critical('The template specified is not a valid one or not found')
            return
        filelist = gen_range(fntemplate, ranges)
    else:
        filelist = gen_tftpfilelist(options.filenamesfile)
    dumpfiles = dict()
    ackbucket = list()
    sportmap = dict()
    start_time = datetime.datetime.now()
    cleansocks = list()    
    finito = False
    lastrecv = time()
    maxlastrecv = 5
    if not os.path.exists(options.dstdir):
        os.mkdir(options.dstdir)
    while 1:
        socklists = list()
        socklists.extend(mydirtysocks.list)
        socklists.extend(cleansocks)
        r,w,e = select.select(socklists,[],[],0.1)
        if len(r) > 0:        
            for s in r:
                recv = s.recvfrom(1024)
                log.debug('Recv: %s' % `recv`)
                lastrecv = time()
                buff = recv[0]
                ipaddr = recv[1]
                try:
                    response = tftpstruct.parse(buff)
                except lib.construct.core.RangeError:
                    response = None
                if response is None:
                    log.warn('Error parsing response')
                    continue
                if response.operation == 'ERROR':
                    if response.data.errorcode == 'FileNotFound':
                        pass
                    else:
                        log.debug(str(response))
                        log.warn( "Time for a siesta.. I think the tftp can't keep up")
                        sleep(2)
                elif response.operation == 'DATA':
                    addresses = (ipaddr,s.getsockname())
                    if addresses not in dumpfiles:
                        dumpfiles[addresses] = dict()                
                    dumpfiles[addresses][response.data.block] = response.data.data
                    if s in mydirtysocks.list:
                        mydirtysocks.list.remove(s)
                        cleansocks.append(s)
                    ackbucket.append([response.data.block,ipaddr,s])            
        else:
            if time() - lastrecv > maxlastrecv:
                log.error('Did not receive a response for %s seconds, quitting' \
                          % (time()-lastrecv))
                finito = True
            if len(ackbucket) > 0:
                block,datadst,s = ackbucket.pop(-1)
                data = t.makeack(block)                
                s.sendto(data,datadst)
            elif finito:
                break
            if not finito:
                try:
                    fn = filelist.next()
                except StopIteration:
                    finito = True
                    continue
                data = t.makerrq(fn)
                log.debug('asking for %s' % fn)
                s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                s.settimeout(5)
                s.sendto(data,dst)
                mydirtysocks.append(s)        
                saddr,sport = s.getsockname()
                sportmap[sport] = fn
            if random.randrange(84) == 42:
                dumpfiles,files = dumpnassemble(dumpfiles,sportmap,options.dstdir,partialdump=False)
                for fn in files:
                    log.info("Dumped %s" % fn)
    
    dumpfiles,files = dumpnassemble(dumpfiles,sportmap,options.dstdir,partialdump=True)
    for fn in files:
        log.info( "pos 2 Dumped %s" % fn)
    log.info( "Total time: %s" % (datetime.datetime.now() - start_time) )

if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG)
    main()